pdns/pdns_recursor.cc

```diff
@@ -396,114 +396,90 @@ static bool isHandlerThread()
-  }
-
-  return LWResult::Result::Success;
-}
+#if 0
+#define TCPLOG(x) cerr << x
+#else
+#define TCPLOG(x)
+#endif
 
 static void TCPIOHandlerWritable(int fd, FDMultiplexer::funcparam_t& var);
 
-LWResult::Result asendtcp(const PacketBuffer& data, TCPIOHandler& handler)
+LWResult::Result asendtcp(const PacketBuffer& data, shared_ptr<TCPIOHandler>& handler)
 {
+  TCPLOG("asendtcp called " << data.size() << endl);
+
   PacketID pident;
-  pident.tcphandler = &handler;
-  pident.tcpsock = handler.getDescriptor();
+  pident.tcphandler = handler;
+  pident.tcpsock = handler->getDescriptor();
   pident.outMSG = data;
 
-  t_fdm->addWriteFD(handler.getDescriptor(), TCPIOHandlerWritable, pident);
+  t_fdm->addWriteFD(handler->getDescriptor(), TCPIOHandlerWritable, pident);
   PacketBuffer packet;
 
   int ret = MT->waitEvent(pident, &packet, g_networkTimeoutMsec);
-  if (ret == 0) { //timeout
-    t_fdm->removeWriteFD(handler.getDescriptor());
+  TCPLOG("asendtcp waitEvent returned " << ret << ' ' << packet.size() << '/' << data.size() << ' ');
+  if (ret == 0) {
+    t_fdm->removeWriteFD(handler->getDescriptor());
+    TCPLOG("timeout" << endl);
     return LWResult::Result::Timeout;
   }
   else if (ret == -1) { // error
-    t_fdm->removeWriteFD(handler.getDescriptor());
+    t_fdm->removeWriteFD(handler->getDescriptor());
+    TCPLOG("PermanentError" << endl);
     return LWResult::Result::PermanentError;
   }
   else if (packet.size() != data.size()) { // main loop tells us what it sent out, or empty in case of an error
+    TCPLOG("PermanentError size mismatch" << endl);
     return LWResult::Result::PermanentError;
   }
 
+  TCPLOG("asendtcp success" << endl);
   return LWResult::Result::Success;
 }
 
-static void handleTCPClientReadable(int fd, FDMultiplexer::funcparam_t& var);
+static void TCPIOHandlerReadable(int fd, FDMultiplexer::funcparam_t& var);
 
-LWResult::Result arecvtcp(PacketBuffer& data, const size_t len, Socket* sock, const bool incompleteOkay)
+LWResult::Result arecvtcp(PacketBuffer& data, const size_t len, shared_ptr<TCPIOHandler>& handler, const bool incompleteOkay)
 {
-  data.clear();
-  PacketID pident;
-  pident.tcpsock=sock->getHandle();
-  pident.inNeeded=len;
-  pident.inIncompleteOkay=incompleteOkay;
-  t_fdm->addReadFD(sock->getHandle(), handleTCPClientReadable, pident);
-
-  int ret = MT->waitEvent(pident, &data, g_networkTimeoutMsec);
-  if (ret == 0) {
-    t_fdm->removeReadFD(sock->getHandle());
-    return LWResult::Result::Timeout;
-  }
-  else if (ret == -1) {
-    t_fdm->removeWriteFD(sock->getHandle());
-    return LWResult::Result::PermanentError;
-  }
-  else if (data.empty()) {// error, EOF or other
-    return LWResult::Result::PermanentError;
+  TCPLOG("arecvtcp called " << len << ' ' << data.size() << endl);
+  size_t pos = 0;
+  data.resize(len);
+  TCPLOG("calling tryRead() " << data.size() << ' ' << len << endl);
+  /* IOState state = */ handler->tryRead(data, pos, len);
```

Comment:
`tryRead()` might throw, for example if the connection has been closed by the remote host, it doesn't look like we handle that?

Reply:
How does this code handle the fact that we might need to write to be able to read, since it does not look at the `IOState`?

Reply:
If tryRead does not return an answer TCPIOHandlerReadbale function. A needRead/needWrite return will be handled there. 

But as I say in the PR intro and also in the `"TCPIOHandlerReadable more? flip to write seems needed???..."` case, I'm not 100% happy they way this is structured, but I fail to see a better way atm.

Maybe the whole approach should be changed to better match the way tryRead/tryWrite work.

Reply:
Doesn't `TCPIOHandlerReadable` get called when the underlying descriptor gets readable? And if so, what happens if it doesn't because we need to write first?

Reply:
All I know is that if I leave out the initial `tryRead()` call in `arecvtcp()` my TLS connection does not work. With (fixed) TCPLOG enabled:
```
...
arecvtcp 1 2 arecvtcp success
arecvtcp called 525 525
arecvtcp 0 0 timeout
```
Maybe what is happening is that the data is already in (because TLS does larger data chunks underneath)? So the fd does not become readable again, but the data is there and we just need to collect it? That is more or less what I'm doing with that initial `tryRead()` call.

Reply:
Yes, it's likely that you need to read first, and then only wait for the descriptor to become readable if you could not read. That's indeed because the data has already been processed at the TLS layer.
But be careful that sometimes reading will require sending data, and 99.9% of the time the socket will be writable so you will not see it, but if the send buffer is full then `tryRead()` will return `IOState::NeedWrite` and you'll need to wait for the descriptor to become writable because it won't become readable otherwise (well, the connection will be closed eventually but that's not what you want).

Reply:
Ok, then what I need to add is a check on the state returnd by the initial `tryWrite()`, if it is `needWrite`, I need to do `addWriteFD` with TCPIOHandlerWritable.

Reply:
Actually I think you need to call `addWriteFD` but with `TCPIOHandlerReadable` as the callback since you really want to read, from your point of view (but be careful about the need to update the mplexer later, though) :)

Reply:
Also, be aware that I only tested TLS on OpenBSD which uses the `pollmultiplexer` so far. There might be subtle differences in the way the multiplexers work...

Reply:
> Actually I think you need to call `addWriteFD` but with `TCPIOHandlerReadable` as the callback since you really want to read, from your point of view (but be careful about the need to update the mplexer later, though) :)

I'm a bit sad that we are forced to either check for either readable or writeable with the current API, while the underlying mechanisms *are* capable of doing both simultaneously...

Reply:
> Hmm, at least in dnsdist we don't really have cases where we want to be notified when the descriptor is either writable OR readable, but we could update the multiplexer code if the recursor needs that.

I don't know yet if we really need it, but atm I think it would make the code easier. But that is just a hunch, no hard data.